consensus/consensus.go

```diff
@@ -306,3 +376,49 @@ func (fc *FastHotStuffConsensus) evictStaleVotesAndTimeouts() {
 		}
 	}
 }
+
+func (fc *FastHotStuffConsensus) storeVote(signaturePayload [32]byte, vote VoteMessage) {
+	votesForBlockHash, ok := fc.votesSeen[signaturePayload]
+	if !ok {
+		votesForBlockHash = make(map[string]VoteMessage)
+		fc.votesSeen[signaturePayload] = votesForBlockHash
+	}
+
+	votesForBlockHash[vote.GetPublicKey().ToString()] = vote
+}
+
+func (fc *FastHotStuffConsensus) hasVotedForView(publicKey *bls.PublicKey, view uint64) bool {
+	// This is an O(n) operation that scales with the number of block hashes that we have stored
+	// votes for. In practice, n will be very small because we evict stale votes, and server.go
+	// will be smart about not processing votes for views we won't be the block proposer for.
+	//
+	// TODO: We can further optimize this by adding a second map[view][publicKey]VoteMessage, but
+	// this is unnecessary for the forseeable future.
+
+	// Compute the string encoding for the public key
+	publicKeyString := publicKey.ToString()
+
+	// Search for the public key's votes across all existing block hashes
+	for _, votesForBlock := range fc.votesSeen {
+		vote, ok := votesForBlock[publicKeyString]
+		if ok && vote.GetView() == view {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (fc *FastHotStuffConsensus) hasTimedOutForView(publicKey *bls.PublicKey, view uint64) bool {
+	timeoutsForView, ok := fc.timeoutsSeen[view]
+	if !ok {
+		return false
+	}
+
+	_, ok = timeoutsForView[publicKey.ToString()]
```

Comment:
The mere existence of a public key in this map indicates we have a valid timeout message, correct?

Reply:
Correct